### DISCUSSION TITLE
Speed up Jekyll builds saving bundle to a directory

### DISCUSSION THREAD

**Discussion by giovannibonetti** [vercel/vercel#3905] [General] [answered]
Our jekyll builds would be way faster if the gem bundle would be cached. Installing all the gems in every build takes a lot of time:

![Screenshot 2020-03-11 18 30 56](https://user-images.githubusercontent.com/3451581/76466502-9a341380-63c6-11ea-9ba5-dff118be9da3.png)

As a reference, here it is how we do it with Google Cloud Build:

- Dockerfile: load `bundle.tar` or `bundle.tar.gz` from the previous build
```Dockerfile
FROM jekyll/jekyll:4

COPY Gemfile* /srv/jekyll/

# Copy bundle cache from the previous build
COPY ./bundle.tar /usr/local/
RUN tar xf /usr/local/bundle.tar -C /usr/local/bundle 2> /dev/null \
    && bundle install --jobs 4 --retry 20 \
    && bundle clean \
    && rm /usr/local/bundle.tar

# Copy app files
COPY . /srv/jekyll/

# Finally build the app
ENV JEKYLL_ENV=production
RUN bundle exec jekyll build
```

- Build script - Since the previous Docker image was already downloaded to help with Docker cache, it can also be used to get `bundle.tar`.
```yml
- id: build production image
  name: gcr.io/cloud-builders/docker
  entrypoint: bash
  args:
  - -c
  - |
    docker create --name ${REPO_NAME}_tmp gcr.io/$ORG_NAME/$REPO_NAME \
      && docker cp ${REPO_NAME}_tmp:/usr/local/bundle . \
      && docker rm -v ${REPO_NAME}_tmp \
      && tar --sort=name --mtime='2025-01-01 00:00Z' -cf ./bundle.tar -C ./bundle . \
      && rm -Rf ./bundle \
      && docker build \
        --cache-from gcr.io/$ORG_NAME/$REPO_NAME \
        -t gcr.io/$ORG_NAME/$REPO_NAME \
        . \
      && rm ./bundle.tar
```

**Comment by styfle** [comment, marked as answer]
Thanks, thats a great idea!

The solution is to add a pattern to [frameworks](https://github.com/zeit/now/blob/b7943e83d278cab4158b7f85ebfd86a8d5a555db/packages/now-static-build/src/frameworks.ts#L416) that matches the path where the gems are installed.

We'll have to first figure out where that path is. Would you like to submit a PR?

**Reply by giovannibonetti** [reply]
> Thanks, thats a great idea!

Thank you!

> We'll have to first figure out where that path is.

When running `bundle install` or just `bundle`, by default the gems are installed at a system-wide folder like `/usr/local/bundle/`. However, when installing them with [`bundle install --deployment`](https://bundler.io/bundle_install.html#deployment) they are installed in the app folder at `./vendor/bundle`. That path can be changed with the `--path` flag.

> Would you like to submit a PR?

Sure. If I understand correctly we would need `cachePattern`, right?

**Reply by giovannibonetti** [reply]
Here it is: #3908

**Reply by mcsdevv** [reply]
Thanks so much @giovannibonetti!
